src/machine/machine_stm32_gpio_reva.go

```diff
@@ -0,0 +1,85 @@
+// +build stm32,!stm32l4,!stm32l5
+
+package machine
+
+import (
+	"device/stm32"
+)
+
+//
+// STM32 allows one interrupt source per pin number, with
+// the same pin number in different ports sharing a single
+// interrupt source (so PA0, PB0, PC0 all share).  Only a
+// single physical pin can be connected to each pin number.
+//
+// To call interrupt callbacks, we record here for each
+// pin number the callback and the actual associated pin.
+//
+
+// Callbacks for pin interrupt events
+var pinCallbacks [16]func(Pin)
+
+// The pin currently associated with interrupt callback
+// for a given slot.
+var interruptPins [16]Pin
+
+// SetInterrupt sets an interrupt to be executed when a particular pin changes
+// state. The pin should already be configured as an input, including a pull up
+// or down if no external pull is provided.
+//
+// This call will replace a previously set callback on this pin. You can pass a
+// nil func to unset the pin change interrupt. If you do so, the change
+// parameter is ignored and can be set to any value (such as 0).
+func (p Pin) SetInterrupt(change PinChange, callback func(Pin)) error {
+	port := uint32(uint8(p) / 16)
+	pin := uint8(p) % 16
+
+	enableEXTIConfigRegisters()
+
+	if callback == nil {
+		stm32.EXTI.IMR.ClearBits(1 << pin)
+		pinCallbacks[pin] = nil
+		return nil
+	}
+
+	if pinCallbacks[pin] != nil {
+		// The pin was already configured.
+		// To properly re-configure a pin, unset it first and set a new
+		// configuration.
+		return ErrNoPinChangeChannel
+	}
+
+	crReg := getEXTIConfigRegister(pin)
+	shift := (pin & 0x3) * 4
+	crReg.ReplaceBits(port, 0xf, shift)
+
+	if (change & PinRising) != 0 {
+		stm32.EXTI.RTSR.SetBits(1 << pin)
+	}
+	if (change & PinFalling) != 0 {
+		stm32.EXTI.FTSR.SetBits(1 << pin)
+	}
+	stm32.EXTI.IMR.SetBits(1 << pin)
+
+	pinCallbacks[pin] = callback
+	interruptPins[pin] = p
+
+	intr := p.registerInterrupt()
+	intr.SetPriority(0)
+	intr.Enable()
+
+	return nil
+}
+
+func handlePinInterrupt(pin uint8) {
+	if stm32.EXTI.PR.HasBits(1 << pin) {
+		// Writing 1 to the pending register clears the
+		// pending flag
+		stm32.EXTI.PR.SetBits(1 << pin)
+
+		callback := pinCallbacks[pin]
+		if callback != nil {
```

Comment:
This is racy, at least in the current implementation. A `func()` value is a doubleword, which is written in two operations. This means that one could be nil while the other isn't. Maybe the `SetInterrupt` method can be written in such a way that the callback is only ever modified when the pin interrupt is disabled? I've done that in other pin interrupt implementations. In that case, you also don't need to check whether the callback is nil (but it's better to do anyway so that the compiler won't insert a nil check).

Reply:
Thanks - will fix

Reply:
Has this been addressed?